koboldcpp: use python3, fix cuda linkage, patch out unused flags, streamline callPackage args

4evy opened the pull request:
## Description of changes

This PR addresses:
https://github.com/NixOS/nixpkgs/pull/314450#issuecomment-2170543384
https://github.com/NixOS/nixpkgs/pull/314450#issuecomment-2208635912

Also updates `koboldcpp` to `1.69.1`

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

4evy committed:
koboldcpp: use python3 instead of python311

Reason: https://github.com/NixOS/nixpkgs/pull/314450#discussion_r1617133115

4evy committed:
koboldcpp: fix CUDA

- Fix CUDA lib linkage
- Add new optional option `arches`

Refs: https://github.com/NixOS/nixpkgs/pull/314450#issuecomment-2170543384
Refs: https://github.com/NixOS/nixpkgs/pull/314450#issuecomment-2208635912
Refs: https://github.com/NixOS/nixpkgs/pull/325540#discussion_r1669900645

Co-authored-by: 百地 希留耶 <65301509+KiruyaMomochi@users.noreply.github.com>
Co-authored-by: mirrexagon <mirrexagon@mirrexagon.com>
Co-authored-by: SomeoneSerge <sergei.kozlukov@aalto.fi>

4evy committed:
koboldcpp: 1.68 -> 1.69.1

Diff: https://github.com/LostRuins/koboldcpp/compare/v1.68...v1.69.1
Changelog: https://github.com/LostRuins/koboldcpp/releases/tag/v1.69.1

DontEatOreo commented:
Hey, @KiruyaMomochi @mirrexagon could you please test this?

DontEatOreo commented:
Result of `nixpkgs-review pr 325540` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>koboldcpp</li>
  </ul>
</details>

DontEatOreo commented:
Result of `nixpkgs-review pr 325540` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>koboldcpp</li>
  </ul>
</details>

maxstrid reviewed (approved):
These changes look good to me

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -28,6 +29,9 @@
   metalSupport ? stdenv.isDarwin && stdenv.isAarch64,
 
+  # You can find a full list here: https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/
+  nvidiaArch ? "",
```
Hi! Please choose the defaults based on `config.cudaSupport` and `config.cudaCapabilities`

SomeoneSerge replied:
Similarly, the `march`/`mtune` could be configured using `stdenv.hostPlatform`. It might not even be necessary to pass this to the build script because the cc-wrapper probably already reflects `hostPlatform`

DontEatOreo replied:
> Similarly, the march/mtune could be configured using stdenv.hostPlatform 

Good catch, I added logic to determine which march to use at eval time.

I also did a bit of research and turns out `mtune` is not used on Apple Silicon but rather it `-mcpu=native` is used instead.

DontEatOreo replied:
> Hi! Please choose the defaults based on `config.cudaSupport` and `config.cudaCapabilities` 

Does Nix know about the host `cudaCapabilities` (As in which one the host is currently using)? 

If yes, it would make it possible to remove this as an option and instead dynamically assign it at `makeFlags`

SomeoneSerge replied:
> Does Nix know about the host cudaCapabilities (As in which one the host is currently using)?

Nix itself doesn't know CUDA exists, but the user is expected to configure Nixpkgs at import time, like `import <nixpkgs. { config.cudaCapabilities = [ .. .]; }`. By default `cudaCapabilities` will be set to the list of all architectures supported by a particular cudatoolkit release for the platform

SomeoneSerge replied:
> If yes, it would make it possible to remove this as an option and instead dynamically assign it at makeFlags

There actually is a plan to extend `cudaPackages.setupCudaHook` to set the relevant flags automatically (like `CMAKE_CUDA_ARCHITECTURES`, etc), but right now downstream derivations are expected to set these ad hoc

DontEatOreo replied:
> There actually is a plan to extend `cudaPackages.setupCudaHook` to set the relevant flags automatically (like `CMAKE_CUDA_ARCHITECTURES`, etc), but right now downstream derivations are expected to set these ad hoc

I guess until that becomes a thing, I could use a hack to expect an array with one string in it, match it with `lib.any` against `cudaPackages.cudaFlags.cudaCapabilities`, and pass it down. I think this would still comply with the `config.cudaCapabilities = [..];` part.

Do you have any suggestions, or is that good enough for now?

SomeoneSerge replied:
> I guess until that becomes a thing, I could use a hack to expect an array with one string in it 

Yeah that's not what usually happens; what we do in other derivations that enforce using a single gencode is we take `buitlins.head cudaCapabilites` and hope that includes PTX

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -43,23 +47,25 @@ let
 
   enableParallelBuilding = true;
 
+  LD_LIBRARY_PATH = [ addDriverRunpath.driverLink ];
```
This is only set at build time, when `${driverLink}/lib` isn't available because of the sandbox.

DontEatOreo replied:
I think I'll need a bit of a helping hand on this here, what exactly do I need to do here?

I prefixed the `LD_LIBRARY_PATH` in `postFixup`, is that enough?

```nix
${lib.optionalString config.cudaSupport ''
  \
    --prefix LD_LIBRARY_PATH: "${lib.makeLibraryPath finalAttrs.LD_LIBRARY_PATH}"
''}
```

SomeoneSerge replied:
`makeLibraryPath` is enough yes, it does add the `/lib` suffix and also chooses the correct output for derivations

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -114,6 +120,7 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     (makeBool "LLAMA_CLBLAST" clblastSupport)
     (makeBool "LLAMA_VULKAN" vulkanSupport)
     (makeBool "LLAMA_METAL" metalSupport)
+    (lib.optionalString (nvidiaArch != "") "CUDA_DOCKER_ARCH=${nvidiaArch}")
```
Code-searching the upstream repo, I don't see any references to the variable other than https://github.com/LostRuins/koboldcpp/blob/7499a6bd4b714aa1bb43284ed87ccf072faa3825/Makefile#L172-L174

Makes me wonder if we should instead set NVCC_APPEND_FLAGS so we can represent a _list_ of gencodes (`cudaPackages.flags.gencodes`).

DontEatOreo replied:
```Make
ifdef CUDA_DOCKER_ARCH
	NVCCFLAGS += -Wno-deprecated-gpu-targets -arch=$(CUDA_DOCKER_ARCH)
else
```

As far as I can tell setting `CUDA_DOCKER_ARCH` will set `NVCCFLAGS` for us, isn't that correct?

DontEatOreo replied:
> cudaPackages.flags.gencodes

I search on nixpkgs (https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20cudaPackages.flags.gencodes&type=code) can couldn't find any reference to `cudaPackages.flags.gencodes` anywhere (Besides this PR)?

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -58,7 +58,7 @@ effectiveStdenv.mkDerivation (finalAttrs: {
   enableParallelBuilding = true;
 
-  LD_LIBRARY_PATH = [ addDriverRunpath.driverLink ];
+  LD_LIBRARY_PATH = [ ] ++ lib.optionals (config.cudaSupport) [ addDriverRunpath.driverLink ];
```
Another note: this will be read by processes at build time (e.g. it'll make gcc look up files in this directory, and cmake, and so on); maybe use a distinguished name, or define something like `makeWrapperArgs`

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
Oh, btw, a heads-up: we'll need commit messages conforming to the manual:

- https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions
- https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions

DontEatOreo replied:
Yeah, no worries, but I'm still working on this PR, so I will rebase it after I'm done working on it

SomeoneSerge replied:
Oh it's also ok for the PR title not to mention every single change but to only provide the gist

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -106,14 +117,28 @@ effectiveStdenv.mkDerivation (finalAttrs: {
+      (lib.optionalString (isx86_64) "-march=x86-64")
+      (lib.optionalString (isAarch64 && isLinux) "-march=armv8.7-a")
+      (lib.optionalString (mtune != "") "-mtune=${mtune}")
+      (lib.optionalString (isDarwin) "-mcpu=native")
```
H'm. We try to avoid `-march=native` and similar flags in Nixpkgs, because then the output isn't determined by the derivation structure (="random", "non-deterministic"); instead, we try to describe the host platform statically (=ideally, produces the same output as it would with `-march=native`, but also it'd produce that even if built on a different machine)

DontEatOreo replied:
I was basing that line on: https://stackoverflow.com/a/72597669 *(Forgot to add a comment about this)*

But there is also: https://stackoverflow.com/a/69926065 so I would go with that instead

SomeoneSerge replied:
Yes looks much better; regarding the `hostPlatform` and the `cc-wrapper`, I'm seeing this:

```console
nix-repl> :lf nixpkgs
Added 15 variables.
nix-repl> legacyPackages.aarch64-darwin.hostPlatform.gcc.cpu
"apple-a13"
```

and this:

https://github.com/NixOS/nixpkgs/blob/6c12901eb4ccb5befc24543a110dc0865d72b62b/pkgs/build-support/cc-wrapper/default.nix#L253-L254

DontEatOreo replied:
Hmmm that's seems weird the `apple-a13` is the iPhone 11 Series CPU...

SomeoneSerge replied:
I would guess that people in https://matrix.to/#/%23macos%3Anixos.org?via=nixos.org&via=matrix.org&via=nixos.dev must know why

4evy committed:
koboldcpp: add postPatch for Darwin

- Add `postPatch` deleting an unused argument for Darwin

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -106,14 +121,31 @@ effectiveStdenv.mkDerivation (finalAttrs: {
+        ;
+    in
+    lib.concatStringsSep " " [
+      (lib.optionalString isx86_64 "-march=x86-64")
```
I'm pretty sure this one is identical to the default behaviour either of the wrapper or of gcc/clang -> can be left out

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -106,14 +119,31 @@ effectiveStdenv.mkDerivation (finalAttrs: {
+    in
+    lib.concatStringsSep " " [
+      (lib.optionalString isx86_64 "-march=x86-64")
+      (lib.optionalString (isAarch64 && isLinux) "-march=armv8.7-a")
```
This one does specify and extended instruction set which will be different from what the user has "requested" with their Nixpkgs configuration (the `config` and `system` they passed to `import <nixpkgs>`). I'm not saying this is necessarily wrong, but at least it's not obvious why it's this way

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -106,14 +119,31 @@ effectiveStdenv.mkDerivation (finalAttrs: {
+    lib.concatStringsSep " " [
+      (lib.optionalString isx86_64 "-march=x86-64")
+      (lib.optionalString (isAarch64 && isLinux) "-march=armv8.7-a")
+      (lib.optionalString (mtune != "") "-mtune=${mtune}")
```
This is set to `""` by default, so it's not actually used in the package set. If the user wishes to reconfigure this, they can use `.override = { stdenv = myCustomStdenv; }` or `.overrideAttrs (oldAttrs: { NIX_CFLAGS_COMPILE = ... ; })`

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -106,14 +119,31 @@ effectiveStdenv.mkDerivation (finalAttrs: {
+      (lib.optionalString (isAarch64 && isLinux) "-march=armv8.7-a")
+      (lib.optionalString (mtune != "") "-mtune=${mtune}")
+      # https://stackoverflow.com/a/69926065
+      (lib.optionalString isDarwin "-mcpu=apple-m1")
```
Similarly, if it's important to provide a build targetting specifically this instruction set maybe we could expose in a custom attribute, but ignoring the hostPlatform configuration is a default that would require some justification. Does the build fail without this flag? Is there a verifiable difference in the performance with and without this flag?

Thanks

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -106,14 +119,31 @@ effectiveStdenv.mkDerivation (finalAttrs: {
+        isDarwin
+        ;
+    in
+    lib.concatStringsSep " " [
+      (lib.optionalString isx86_64 "-march=x86-64")
```
Could have been `concatStringsSep " " (optionals ...)`

DontEatOreo commented:
I decided to revert `march` back to being an optional setting since we can't set it manually

After doing some testing, it seems like `env.NIX_CFLAGS_COMPILE` wasn't being respected *(as when building koboldcpp on x86_64-linux, it was using native regardless of any custom `march` or `mtune`)*

So to deal with that we'll need to use `postPatch` to manually patch the `Makefile`

SomeoneSerge commented:
> So to deal with that we'll need to use postPatch to manually patch the Makefile

Oh, I see. For the llama part, you can set `cmakeFlags = [ (lib.cmakeBool "LLAMA_NATIVE" false) ]`, but yes, seems like you won't be able to avoid some patching. Not having those removed can be considered a bug on nixpkgs side, but it's ok to handle that in a separate PR if you'd rather

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -28,38 +29,53 @@
 
+  # You can find a full list here: https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/
+  # For example if you're on an GTX 1080 that means you're using "Pascal" and you need to pass "6.0"
+  cudaCapabilities ? cudaPackages.cudaFlags.cudaCapabilities or [ ],
```
`or [ ]` is unnecessary, we guarantee this attribute to exist; that said seems like what you end up using is more like

```suggestion
cudaCapabilities ? lib.take 1 cudaPackages.flags.realArches,
```

E.g.

```
❯ nix repl -f '<nixpkgs>'
...
nix-repl> lib.take 1 cudaPackages.flags.realArches
[ "sm_60" ]
```

However, are you sure you want to use `sm_` and not `compute_`?

DontEatOreo replied:
Looking at: https://github.com/LostRuins/koboldcpp/blob/7499a6bd4b714aa1bb43284ed87ccf072faa3825/examples/llava/MobileVLM-README.md?plain=1#L197

It it seems like that's the way to do it? I will go with `builtins.head cudaPackages.cudaFlags.realArches`

SomeoneSerge replied:
Sure, but to be clear, this will be (unless upstream sneaks in some more flags in the buid script) only compatible with a single generation of GPUs, i.e. I think this will leave out the PTX code which could be jit-ed at runtime for other/newer GPUs

DontEatOreo replied:
I'm a bit unsure of what I can do here. We can't grab the actual `arch` the host has, so this only really leaves us with the choice to grab the lowest one...

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -28,38 +29,53 @@
+    --prefix LD_LIBRARY_PATH: "${lib.makeLibraryPath [ addDriverRunpath.driverLink ]}"
+  '';
+
+  containsValidCapability = lib.any (
+    x: x == builtins.head cudaCapabilities
+  ) cudaPackages.cudaFlags.cudaCapabilities;
```
I think it's ok to leave this logic out; the default you've set after `cudaCapabilities ? ...` is guaranteed to be correct, and if the package is called with different args it must be that the user supplied them in an override; the user probably knows better what they want

DontEatOreo replied:
Yeah that's true, I will remove this

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -133,10 +149,26 @@ effectiveStdenv.mkDerivation (finalAttrs: {
+  # Remove `warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]` on Darwin
+  postPatch = ''
+    substituteInPlace Makefile \
+      --replace-fail '		CFLAGS += -march=native -mtune=native' '		CFLAGS += -march=${
+        (if (march != "") then "${march}" else "native")
+      } -mtune=${(if (mtune != "") then "${mtune}" else "native")}' \
```
Wouldn't this get the job done?

```suggestion
      --replace-fail "-march=native" "" \
      --replace-fail "-mcpu=native" "" \
      --replace-fail "-mtune=native" "" \
```

DontEatOreo replied:
How so? It just removes those flags rather than replacing them with custom one

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -133,10 +149,26 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     runHook postInstall
   '';
 
+  # `env.NIX_CFLAGS_COMPILE` is not respected by KoboldCpp, so the Makefile has to be manually patched
```
`NIX_CFLAGS_COMPILE` is always respected because it's interpreted by the cc-wrapper; the reason the flags didn't have any effect is because the Makefile adds flags, and the cc-wrapper adds the same flags with different values, and these are all passed on to gcc which is free to choose which ones to respect

By removing the conflicting flags from the Makefile entirely, we delegate the configuration job to the cc-wrapper and all that smart nixpgks tooling which I mostly don't understand

DontEatOreo replied:
Thanks for the clarification but at the same time in `Makefile` there is only this part about `-mcpu`

https://github.com/LostRuins/koboldcpp/blob/cc133401db7d3271ff3907d7c79e4bf99d25ae83/Makefile#L302-L308

Which cannot get triggered on any Darwin machine yet the `-mcpu` flag gets ignored in `NIX_CFLAGS_COMPILE`, so I'm not too sure just removing stuff from `Makefile` will do the trick, at least on macOS since it uses `clang` to build

SomeoneSerge replied:
H'm. Maybe we step back: how do you know `mcpu` is ignored?

> it uses clang to build

Clang is used through the cc-wrapper as well

DontEatOreo replied:
I'm not on the PC anymore *(I will edit this comment later today with a log)* but it will absent in the CFLAGS print at the start of the build.

DontEatOreo replied:
Here is the log of `koboldcpp` build on `aarch64-darwin`:

```console
Running phase: buildPhase
build flags: -j8 SHELL=/nix/store/s30jrpgav677fpc9yvkqsib70xfmx7xi-bash-5.2p26/bin/bash LLAMA_VULKAN=1 LLAMA_METAL=1
I llama.cpp build info:
I UNAME_S:  Darwin
I UNAME_P:  arm
I UNAME_M:  arm64
I CFLAGS:   -I. -Iggml/include -Iggml/src -Iinclude -Isrc -I./include -I./include/CL -I./otherarch -I./otherarch/tools -I./otherarch/sdcpp -I./otherarch/sdcpp/thirdparty -I./include/vulkan -O3 -fno-finite-math-only -DNDEBUG -std=c11   -fPIC -DLOG_DISABLE_LOGS -D_GNU_SOURCE -DGGML_USE_LLAMAFILE -pthread -s -Wno-deprecated -Wno-deprecated-declarations -pthread -DGGML_USE_ACCELERATE -DGGML_USE_BLAS -DGGML_USE_METAL -DGGML_METAL_NDEBUG -DSD_USE_METAL
I CXXFLAGS: -I. -Iggml/include -Iggml/src -Iinclude -Isrc -I./common -I./include -I./include/CL -I./otherarch -I./otherarch/tools -I./otherarch/sdcpp -I./otherarch/sdcpp/thirdparty -I./include/vulkan -O3 -fno-finite-math-only -DNDEBUG -std=c++11 -fPIC -DLOG_DISABLE_LOGS -D_GNU_SOURCE -DGGML_USE_LLAMAFILE -pthread -s -Wno-multichar -Wno-write-strings -Wno-deprecated -Wno-deprecated-declarations -pthread -DGGML_USE_ACCELERATE -DGGML_USE_BLAS -DGGML_USE_METAL -DSD_USE_METAL
I LDFLAGS:  -framework Accelerate -framework Foundation -framework Metal -framework MetalKit -framework MetalPerformanceShaders
I CC:       clang version 16.0.6
I CXX:      clang version 16.0.6
```

This is when I'm using `env.NIX_CFLAGS_COMPILE` without `postPatch`. I had the mcpu set to `apple-m1`, but as you can see, it's absent.

DontEatOreo replied:
> `EDIT: Could this have been the effect of LLAMA_NATIVE=ON which sets march=native?`

Hmm? That's in the `CMakeLists.txt` we're using `Makefile`

SomeoneSerge replied:
>  I had the mcpu set to apple-m1, but as you can see, it's absent.

Oh, these must just be the flags passed by the Makefile. The makefile passes these flags to cc-wrapper, and the cc-wrapper passes them to gcc/clang together with NIX_CFLAGS_COMPILE. I suppose you could inspect the final binary to see what kind of instructions it uses

DontEatOreo replied:
After setting the `env.NIX_DEBUG = 1;` I was able to spot the `apple-m1` flags. I suppose in this case, I will implement the `postPatch` that just simply deletes the `march`, `mtune`, and `mcpu` so they can be defined by Nix

DontEatOreo commented:
I think at this point the PR is done.

@SomeoneSerge I would like to thank you a lot for this detailed review. This PR has helped me learn a lot, and I appreciate all the help you gave!

DontEatOreo commented:
Result of `nixpkgs-review pr 325540` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>koboldcpp</li>
  </ul>
</details>

DontEatOreo commented:
Result of `nixpkgs-review pr 325540` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>koboldcpp</li>
  </ul>
</details>

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -133,10 +148,20 @@ effectiveStdenv.mkDerivation (finalAttrs: {
+  # Remove `warning: argument unused during compilation: '-s' [-Wunused-command-line-argument]` on Darwin
+  postPatch = ''
+    substituteInPlace Makefile \
+      --replace-warn "-march=native" "" \
+      --replace-warn "-mtune=native" "" \
+      --replace-warn " -s " " "
```
TIL: https://github.com/NixOS/nixpkgs/blob/4b6a0890371976e066083789f3039c6f78af95d6/pkgs/build-support/cc-wrapper/cc-wrapper.sh#L121-L134

But maybe it doesn't hurt to keep this

DontEatOreo replied:
That's really handy. Well, I was looking at the log with `NIX_DEBUG`, and indeed it removes `native` automatically. So, patching the Makefile for that is pointless, although it is confusing since the print from KoboldCpp is misleading.

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -28,38 +29,48 @@
   # For ARM here: https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html
   # If you set "march" to "native", specify "mtune" as well; otherwise, it will be set to "generic". (credit to: https://lemire.me/blog/2018/07/25/it-is-more-complicated-than-i-thought-mtune-march-in-gcc/)
-  # For example, if you have an AMD Ryzen CPU, you will set "march" to "x86-64" and "mtune" to "znver2"
   march ? "",
+  # Apple Silicon Chips (M1, M2, M3 and so on DO NOT use mtune)
+  # For example, if you have an AMD Ryzen CPU, you will set "mtune" to "znver2"
   mtune ? "",
+  mcpu ? "",
```
I'm still not convinced it's worth introducing the callPackage arguments. Conceptually, it'd be more consistent to require users to use something like:

```nix
let
  withMarch = march: oldAttrs: {
    env = oldAttrs.env // { NIX_CFLAGS_COMPILE = env.NIX_CFLAGS_COMPILE or "" + "-march=${march}"; };
  };
  pkgs' = mapAttrs (_: withMarch "x86_64-v3") pkgs;
in {
  inherit (pkgs') koboldcpp llama-cpp ollama;
}
```

It's a bit of a mouthful, but it's consistent and (fairly) universal. To contrast, while you could `koboldcpp.override { march = "..."; }`, you couldn't `blender.override { march = "..."; }` because it's an ad hoc argument.

I guess it wouldn't hurt much to let this through, I'll look at this again in the morning. I'll try not to delay merging any further than that

DontEatOreo replied:
That's a good point, thanks for bringing this up. I guess it's more appropriate to let users "merge/append" `march`, `mtune`, and `mcpu` to `env` themselves. I was mainly introducing these options because they were easier from a users POV, but since other AI packages already have this convention, it would be nice on my part to follow it too.

4evy committed:
koboldcpp: remove march, mtune and mtune

After going over a detailed conversation with @SomeoneSerge about
`cc-warpper`, `march` `mtune` and `mcpu`, I've come to the conclusion that
it is unnecessary to have those flags as ad hoc options. Instead, it's
best for users to "merge" their own `env` on top of the current one

Refs: https://github.com/NixOS/nixpkgs/pull/325540#discussion_r1674917673
Refs: https://github.com/NixOS/nixpkgs/pull/325540#discussion_r1676059901
Refs: https://github.com/NixOS/nixpkgs/pull/325540#discussion_r1676556732
Refs: https://github.com/NixOS/nixpkgs/pull/325540#discussion_r1676560445

Co-authored-by: SomeoneSerge <sergei.kozlukov@aalto.fi>

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -105,15 +106,13 @@ effectiveStdenv.mkDerivation (finalAttrs: {
     (makeBool "LLAMA_CLBLAST" clblastSupport)
     (makeBool "LLAMA_VULKAN" vulkanSupport)
     (makeBool "LLAMA_METAL" metalSupport)
+    (lib.optionalString cublasSupport "CUDA_DOCKER_ARCH=sm_${builtins.head arches}")
```
FYI: This adds two extra spaces onto the command line; `++ optionals cublasSupport [ "..." ]` doesn't

SomeoneSerge commented on pkgs/by-name/ko/koboldcpp/package.nix:
```diff
@@ -18,6 +19,9 @@
   cublasSupport ? config.cudaSupport,
+  # You can find a full list here: https://arnon.dk/matching-sm-architectures-arch-and-gencode-for-various-nvidia-cards/
+  # For example if you're on an GTX 1080 that means you're using "Pascal" and you need to pass "sm_60"
+  arches ? cudaPackages.cudaFlags.arches or [ ],
```
```suggestion
  cudaArches ? cudaPackages.flags.arches,
```

- A more specific name might be preferable if 1) one was to add rocm support, 2) one were to add a package named "arches" (I hope this never happens, that's a bad name)
- The `or [ ]` is unnecessary, at this point we'd be afraid to change the interface of `cudaPackages.flags`

But IMO this is already good to go

SomeoneSerge reviewed (approved):
Thank you @DontEatOreo!

I hope you don't mind, I'll change the PR title before the merge, to have the `<pkg>: ...` prefix as before
